ipld/src/selector/walk.rs

```diff
@@ -0,0 +1,208 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::super::{Error, Ipld, Path, PathSegment};
+use super::Selector;
+use async_recursion::async_recursion;
+use async_trait::async_trait;
+use cid::Cid;
+
+impl Selector {
+    /// Walks all nodes visited (not just matched nodes) and executes callback with progress and
+    /// Ipld node. An optional link loader/ resolver is passed in to be able to traverse links.
+    pub async fn walk_all<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
+        callback: F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld, VisitReason) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        Progress {
+            resolver,
+            path: Path::default(),
+        }
+        .walk_all(ipld, self, &callback)
+        .await
+    }
+
+    /// Walks a graph of Ipld nodes, executing the callback only on the nodes "matched".
+    /// If a resolver is passed in, links will be able to be traversed.
+    pub async fn walk_matching<L, F>(
+        self,
+        ipld: &Ipld,
+        resolver: Option<L>,
+        callback: F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        self.walk_all(ipld, resolver, |prog, ipld, reason| -> Result<(), String> {
+            if let VisitReason::SelectionMatch = reason {
+                return callback(prog, ipld);
+            }
+            Ok(())
```

Comment:
```suggestion
            match reason {
                VisitReason::SelectionMatch => callback(prog, ipld),
                VisitReason::SelectionCandidate => Ok(()),
            }
```
Let me know what you think about this, I'm a sucker for exhaustiveness and symmetry myself – and Rust kinda discourages early returns with how you can omit the `return` keyword

Reply:
idk thoughts were it's more readable, I couldn't care less though so I'll just change. It's obviously functionally the same